File: webapp/src/domain/workoutBuilder.ts
import type { WorkoutSession } from "./models";

/**
 * In-browser fallback for the workout_coach capability, ported from the
 * backend's `build_session`. Never programs through pain; rest is a valid
 * choice.
 */
export function buildSession(
  goal = "general fitness",
  minutes = 30,
  equipment = "none",
): WorkoutSession {
  const total = Math.min(90, Math.max(10, Math.trunc(minutes) || 30));
  const warmup = Math.max(3, Math.trunc(total / 6));
  const cooldown = Math.max(3, Math.trunc(total / 8));
  const main = total - warmup - cooldown;

  let block: string[];
  const g = goal.toLowerCase();
  if (g.includes("strength")) {
    block = ["squats", "push-ups", "rows", "glute bridges", "plank"];
  } else if (g.includes("cardio")) {
    block = ["brisk intervals", "step-ups", "mountain climbers", "easy jog"];
  } else {
    block = ["bodyweight circuit", "core", "mobility flow"];
  }

  return {
    goal,
    totalMinutes: total,
    warmupMin: warmup,
    mainMin: main,
    cooldownMin: cooldown,
    mainBlock: block,
    equipment,
    note: "Stop if anything hurts. A rest day is a valid choice, not a broken streak.",
    timestampMillis: Date.now(),
    fromFallback: true,
    personaReaction: "",
  };
}

File: android/app/src/main/java/com/judgemycal/app/domain/WorkoutBuilder.kt
package com.judgemycal.app.domain

/**
 * On-device fallback for the workout_coach capability, ported from the
 * backend's `build_session`. Never programs through pain, never rewards
 * overtraining; rest is a valid choice.
 */
object WorkoutBuilder {

    fun buildSession(
        goal: String = "general fitness",
        minutes: Int = 30,
        equipment: String = "none",
    ): WorkoutSession {
        val total = minutes.coerceIn(10, 90)
        val warmup = maxOf(3, total / 6)
        val cooldown = maxOf(3, total / 8)
        val main = total - warmup - cooldown

        val block = when {
            goal.contains("strength", ignoreCase = true) ->
                listOf("squats", "push-ups", "rows", "glute bridges", "plank")
            goal.contains("cardio", ignoreCase = true) ->
                listOf("brisk intervals", "step-ups", "mountain climbers", "easy jog")
            else -> listOf("bodyweight circuit", "core", "mobility flow")
        }

        return WorkoutSession(
            goal = goal,
            totalMinutes = total,
            warmupMin = warmup,
            mainMin = main,
            cooldownMin = cooldown,
            mainBlock = block,
            equipment = equipment,
            note = "Stop if anything hurts. A rest day is a valid choice, not a broken streak.",
            timestampMillis = System.currentTimeMillis(),
            fromFallback = true,
        )
    }
}
